### ISSUE TITLE
Allow relative min_delta for EarlyStopping

### ISSUE THREAD

**Issue by phemmer** [tensorflow/tensorflow#35970]
**System information**
- TensorFlow version (you are using): 2.1.0
- Are you willing to contribute it (Yes/No): maybe

**Describe the feature and the current behavior/state.**
Currently the [EarlyStopping callback](https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/EarlyStopping) `min_delta` is an absolute value. Meaning that the metric being monitored must improve by that fixed amount. The proposal is to add the ability to use a relative threshold instead (e.g. a ratio or percentage).

An example use case is probably easier to understand. Lets say we're doing a hyperparameter search, so we're running a bunch of different models, and we want to stop early on the bad ones. On good models our loss value is in the single digits. So we set the parameters `baseline=50` (a little high so that we allow models to start off somewhat bad), and `min_delta=0.1`. Now with this, if a model gets under 50, but drops to just 48.9, that'll keep it from triggering the abort threshold. If this were instead a ratio, such that `min_delta=0.1` means `10%`, the loss would have to go from 50 to 45 to keep it from triggering, and a good model at say 1.2 would only have to drop to 1.08.

Now we could try other approaches to address this problem, such as lowering `baseline`, and increasing `patience` (allowing bad starts a little longer to come down further), but by increasing patience, this increases the amount of time we have to wait for good models to stop as well.

&nbsp;

Instead of adding a flag for toggling whether `min_delta` is absolute or relative, I think it might be better to allow the user to provide their own "is this new value good enough" function/lambda. Currently you can technically override `EarlyStopping.monitor_op`, but it's clearly not designed to operate that way as the arguments to `monitor_op` aren't simple "best" and "current" values, and there's also code which directly checks whether `monitor_op == np.less`, which overriding breaks.

So ultimately the request is to either add a new argument to the initializer accepting the signature `(current:float, best:float) -> bool`, or adjust the code to allow overriding `monitor_op` with that signature.

**Will this change the current api? How?**
Maybe, if the direction is to add an additional parameter to the initializer.

**Who will benefit with this feature?**
Those doing hyperparameter searches, or potentially other cases.

**Any Other info.**

**Comment by phemmer** [comment]
Here's an actual real world example:
![image](https://user-images.githubusercontent.com/1826947/72591154-7627ea80-38cd-11ea-98cf-3bd1dcc6f4b9.png)
This is the tensorboard output for several different models being tested. The ones worth investigating further are the ones that are linear on the graph (linear even though the graph is logarithmically scaled), as they show that with further training, they will likely continue to improve. This is exactly what a relative `min_delta` would help find.
I want to have `min_delta` set low enough so that the ones like that orange at the bottom continue all the way to the end. But by setting `min_delta` low, the ones that continued to the end, but were virtually flat were also allowed to continue.
In addition, some of the ones that were terminated early were very linear (again linear on the logarithmic scale), and if allowed to continue might have ended up among the best results.

**Comment by ducvinh9** [comment]
I hope someone can implement this idea : relative _min_delta_. It is actually the standard convergence criterion in optimization. A slow drop doesn't mean convergence at all.

**Comment by leleogere** [comment]
I agree that a relative `min_delta` is indispensable for hyperparameter searching.
